Hgc l1clustering rebased pre5 tmp

mastrolorenzo opened the pull request:
Update to the clusters classes in the Dataformat and in the clustering function in the implementation classes. Adding of a simple 2D-clustering algorithm based on dR(x,y) + a version of a 3D-clustering used to cross-check the 2D-clustering. Modification of the calibration class, adding a function to calibrate from hwPt in transverse mip.

mastrolorenzo committed:
this is the version of the code using pointers in the context of the clustering and multiclutering next commit will use references

mastrolorenzo committed:
this is the version of the code using pointers in the context of the clustering and multiclutering next commit will use references

mastrolorenzo committed:
modifications to the cluster implementation functions in order to take collection references instead of unique_ptr to collections + add of parameter for geometrical cluster 2D in hgcalTriggerPrimitiveDigiProducer_cfi.py

mastrolorenzo committed:
modifications to the cluster implementation functions in order to take collection references instead of unique_ptr to collections + add of parameter for geometrical cluster 2D in hgcalTriggerPrimitiveDigiProducer_cfi.py

mastrolorenzo committed:
c2d clustering algorithm core is compiling: defined addTC() and dist() member functions as const in HGCalCluster class

mastrolorenzo committed:
c2d clustering algorithm core is compiling: defined addTC() and dist() member functions as const in HGCalCluster class

mastrolorenzo committed:
2D cluster is ready to be tested

mastrolorenzo committed:
2D cluster is ready to be tested

mastrolorenzo committed:
small modification to the 2D-clustering algorithm + 3D-clustering as in IC code just to debug the C2d (need the sum of the energy)

mastrolorenzo committed:
small modification to the 2D-clustering algorithm + 3D-clustering as in IC code just to debug the C2d (need the sum of the energy)

mastrolorenzo committed:
2D cluster algorithm implemented (simple 2D-dR(x,y) approach)
+ modifications to the cluster classes and their implementation:
parsing of the eventSetup and Event in the clusterise functions
(needed to retrieve GlobalPointers to the cartesian position of the 2D) + creation of a *temporary-C3d* used to test the C2d + global cleaning

mastrolorenzo committed:
2D cluster algorithm implemented (simple 2D-dR(x,y) approach)
+ modifications to the cluster classes and their implementation:
parsing of the eventSetup and Event in the clusterise functions
(needed to retrieve GlobalPointers to the cartesian position of the 2D) + creation of a *temporary-C3d* used to test the C2d + global cleaning

mastrolorenzo committed:
prepare for rebase

mastrolorenzo committed:
rebased: compile and running

mastrolorenzo committed:
Merged hgc_l1clustering_rebased from repository mastrolorenzo

mastrolorenzo committed:
adjust of the calibration class to host a new member function that returns the calibrated pT in mip units

mastrolorenzo committed:
Cluster classes and implementation modified. No more mutable members.

mastrolorenzo committed:
cleaning

mastrolorenzo committed:
cluster and multicluster class modified: methods that return an std::vector<HGCalTriggerCell*> and std::vector<Cluster*> have been implemented + removal of Event and EventSetup from the clustering functions parameters, using the GlobalPointer from TC

mastrolorenzo committed:
update the rebased_on_pre5 branch with what has been done on 03/03/17 in hgc_l1clustering_rebased

jbsauvan commented on DataFormats/L1THGCal/BuildFile.xml:
```diff
@@ -1,3 +1,4 @@
+<use   name="RecoLocalCalo/HGCalRecAlgos"/>
```
Is this needed? I don't see any dependency on this package in DataFormats/L1THGCal.

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
 #ifndef DataFormats_L1Trigger_HGCalCluster_h
 #define DataFormats_L1Trigger_HGCalCluster_h
 
-#include "DataFormats/L1Trigger/interface/L1Candidate.h"
```
HGCalCluster inherits from L1Candidate. So this include should be kept.

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
-          );
-
-      ~HGCalCluster();
+#include "DataFormats/DetId/interface/DetId.h"
```
Is this needed? I don't see any DetId in the class.
There are HGCalDetId's in the .cc file, so better to have an HGCalDetId include in the .cc file instead.

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
-      void setModule  (uint32_t module) {module_ = module;}
-      void setHOverE  (uint32_t hOverE){hOverE_ = hOverE;}
+#include "Math/Vector3D.h"
+#include "TMath.h"
```
From what I see this is included for Sqrt(). I tend to prefer the usage of standard libraries instead (std::sqrt()).
Also this is not needed in the .h, but only in the .cc.

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
+        ~HGCalCluster();
+
+        /* trigger-cell collection pertinent to the cluster*/
+        BXVector<const l1t::HGCalTriggerCell*>  tcs() const { return tcs_; }        
```
Several things here:
1. The raw pointers to the trigger cells will probably be broken outside the class where they are produced. Did you check that? Instead it's probably better to use persistent pointers, e.g. edm::PtrVector
2. You may want to avoid copying the vector and return a const reference instead
3. The function should be clear just by looking at its name, in particular for public functions. For instance `triggerCells()` would be much better than `tcs()`

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
+        BXVector<const l1t::HGCalTriggerCell*>  tcs() const { return tcs_; }        
+
+        /* helpers */
+        bool isPertinent( const l1t::HGCalTriggerCell &tc, double dR ) const;
```
The HGCalCluster class should be as generic as possible. For instance, this is not necessarily a 2D cluster. The implementation of `isPertinent()` looks very specific. In my opinion this definition of 'pertinent' should not belong here, it's related to the clustering algorithm, not to the data format.

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
+
+        /* helpers */
+        bool isPertinent( const l1t::HGCalTriggerCell &tc, double dR ) const;
+        void addTC( const l1t::HGCalTriggerCell &tc );
```
The name can be improved. For instance `addTriggerCell` instead of `addTC`.

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
+
+        /* get info */
+        bool isValid()      const { return true;  }
+        uint32_t hwPt()     const { return hwPt_; }
```
Why redefining `hwPt`? This is already a member of the `L1Candidate` class.

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
+        uint32_t hwPt()     const { return hwPt_; }
+        double mipPt()      const { return mipPt_; }
+        //uint32_t hwSeedPt() const { return hwSeedPt_; }
+        double dist( const l1t::HGCalTriggerCell &tc ) const; /* return distance in 'cm' */
```
Better name: `distance` instead of `dist`

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
+        //uint32_t hwSeedPt() const { return hwSeedPt_; }
+        double dist( const l1t::HGCalTriggerCell &tc ) const; /* return distance in 'cm' */
+        
+        ROOT::Math::XYZVector centre() const { return centre_; }
```
1. For consistency with other parts of the code (geometry, trigger cell), it would be better to use `GlobalPoint` for positions
2. Better to return a const reference instead of a copy of the object

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
+        double dist( const l1t::HGCalTriggerCell &tc ) const; /* return distance in 'cm' */
+        
+        ROOT::Math::XYZVector centre() const { return centre_; }
+        ROOT::Math::XYZVector centreNorm() const { return centre_/centre_.z(); }
```
1. For consistency with other parts of the code (geometry, trigger cell), it would be better to use `GlobalPoint` for positions
2. Better to return a const reference instead of a copy of the object
3. This doesn't look like the normalized vector, or the norm of the vector. What does 'Norm' means?

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
+        int32_t zside()     const;
+        uint32_t module()    const { return module_; }
+
+        ClusterShapes shapes; /* ??? */
```
Why the questions marks?

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
 #ifndef DataFormats_L1Trigger_HGCalCluster_h
```
It's seems that all the indentations in this file have been changed. It makes the diff much more difficult to see. Please keep the indentation of the original author.

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
-  typedef BXVector<HGCalCluster> HGCalClusterBxCollection;
+        
+        
+        BXVector<const l1t::HGCalTriggerCell*> tcs_;
```
Same comment as above concerning raw pointers and PtrVector

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
+        uint32_t seedDetId_;
+ 
+        /* Centre weighted with energy */
+        ROOT::Math::XYZVector centre_;
```
Same as above: better to use `GlobalPoint` for consistency

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
+        ROOT::Math::XYZVector centre_;
+
+        /* Energies */
+        uint32_t hwPt_;
```
This variable already exists in `L1Candidate`

jbsauvan reviewed (changes requested):
Hi @mastrolorenzo 
Just a first part of the comments. More will arrive later.

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,74 +1,77 @@
 #ifndef DataFormats_L1Trigger_HGCalMulticluster_h
 #define DataFormats_L1Trigger_HGCalMulticluster_h
 
-#include "DataFormats/L1Trigger/interface/L1Candidate.h"
```
Same as for HGCalCluster. HGCalMultiCluster derives from L1Candidate, so the include should be kept.

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,74 +1,77 @@
-        
+
+        /* cluster collection pertinent to the multicluster*/
+        BXVector<const l1t::HGCalCluster*> clusters() const { return clusters_; }        
```
Same comments as for HGCalCluster.
Also, the same feature has been implemented for reco multi clusters, so why reinventing something different?

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,74 +1,77 @@
+        BXVector<const l1t::HGCalCluster*> clusters() const { return clusters_; }        
+
+        /* helpers */
+        bool isPertinent( const l1t::HGCalCluster & clu, double dR ) const;
```
Same comment as for HGCalCluster. This is a function related to the algorithm, not the data format.

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,74 +1,77 @@
+
+        /* helpers */
+        bool isPertinent( const l1t::HGCalCluster & clu, double dR ) const;
+        void addClu( const l1t::HGCalCluster & clu );
```
`addCluster()` would be a clearer name than `addClu()`

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,74 +1,77 @@
-        uint32_t nTotLayer()  const {return nTotLayer_;}
-        
-        uint32_t hOverE() const {return hOverE_;}
+        ROOT::Math::XYZVector centre() const { return centre_; } /* in normal plane (x, y, z)*/
```
Same comment as for HGCalCluster. Would be better to use `GlobalPoint` for consistency with the rest of the code.

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,74 +1,77 @@
-        
-        uint32_t hOverE() const {return hOverE_;}
+        ROOT::Math::XYZVector centre() const { return centre_; } /* in normal plane (x, y, z)*/
+        ROOT::Math::XYZVector centreNorm() const { return centreNorm_; } /* in normalized plane (x/z, y/z, z/z)*/
```
ditto

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,74 +1,77 @@
+        ROOT::Math::XYZVector centre() const { return centre_; } /* in normal plane (x, y, z)*/
+        ROOT::Math::XYZVector centreNorm() const { return centreNorm_; } /* in normalized plane (x/z, y/z, z/z)*/
+
+        uint32_t hwPt() const { return hwPt_; }
```
hwPt already exists in L1Candidate

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,74 +1,77 @@
+        uint32_t hwPt() const { return hwPt_; }
+        double mipPt() const { return mipPt_; }
+
+        uint32_t nTotLayer()  const { return nTotLayer_; } /* not working */
```
What do you mean by 'not working'?

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,74 +1,77 @@
-        uint32_t hwPtHad_;
-        uint32_t hwSeedPt_;
-        
+        BXVector<const l1t::HGCalCluster*> clusters_;
```
Same comment as above and HGCalCluster

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,74 +1,77 @@
+        BXVector<const l1t::HGCalCluster*> clusters_;
+
+        /* Energies */
+        uint32_t hwPt_;
```
Already existing in L1Candidate

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,74 +1,77 @@
+        double mipPt_;
+
+        /* centre in norm plane */
+        ROOT::Math::XYZVector centre_;
```
Same comment as above

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,74 +1,77 @@
+        ROOT::Math::XYZVector centre_;
+
+        /* barycentre */
+        ROOT::Math::XYZVector centreNorm_;
```
Ditto

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,74 +1,77 @@
         // HGC specific information
-        uint32_t subDet_;
+        /* detId of the first cluster in the multicluster */
+        uint32_t firstCluId_;
```
could you find a slightly better name? `firstClusterId_`? `seedClusterId_`?

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,74 +1,77 @@
+        /* detId of the first cluster in the multicluster */
+        uint32_t firstCluId_;
+
         uint32_t nTotLayer_;
```
What is this variable `nTotLayer_`?

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,74 +1,77 @@
         // identification variables
         uint32_t hOverE_; 
+
+        int32_t zside_;
```
Can you instead use the detID of the first cluster to derive zside? To be consistent with what is done in HGCalCluster with the seedId.

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalTriggerCell.h:
```diff
@@ -5,41 +5,57 @@
-            uint32_t detid_;
-            GlobalPoint position_;
+        
+    public:
```
Please don't change all the indentations of the original file. It is a pain to see the real diff.

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalTriggerCell.h:
```diff
@@ -5,41 +5,57 @@
+        uint32_t detId() const {return detid_;}
+        const GlobalPoint& position() const { return position_; }
+        
+        int zside() {                
```
This should be a `const` method (it doesn't modify the object)

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalTriggerCell.h:
```diff
@@ -5,41 +5,57 @@
+            HGCalDetId trgdetid(detid_);                
+            return trgdetid.zside();               
+        }
+        int layer() {                
```
Ditto: This should be a `const` method

jbsauvan commented on DataFormats/L1THGCal/src/HGCalCluster.cc:
```diff
@@ -3,32 +3,144 @@
+{
+
+    if( tcs_.size() == 0 ) 
+        seedDetId_ = tc.detId();
```
This is dangerous to put the statement in a new line without brackets. Please either
-  put the statement on the same line as the condition
```cpp
if( tcs_.size() == 0 ) seedDetId_ = tc.detId();
```
- or put brackets:
```cpp
if( tcs_.size() == 0 ) {
  seedDetId_ = tc.detId();
}
```

jbsauvan commented on DataFormats/L1THGCal/src/HGCalCluster.cc:
```diff
@@ -3,32 +3,144 @@
+                                      tc.position().y(), 
+                                      tc.position().z() );
+
+    return TMath::Sqrt( ( tcPointXYZ - centre_ ).Mag2() );
```
I would prefer to use `std::sqrt()` from the standard library

jbsauvan commented on DataFormats/L1THGCal/src/HGCalCluster.cc:
```diff
@@ -3,32 +3,144 @@
     }
-  }
-  return res;
+    else if( mipPt() == cl.mipPt() ) {
```
`mipPt` is a double. All the following is not needed.

jbsauvan commented on DataFormats/L1THGCal/src/HGCalMulticluster.cc:
```diff
@@ -2,35 +2,76 @@
 
 using namespace l1t;
 
-HGCalMulticluster::HGCalMulticluster( const LorentzVector p4, 
```
Where did this constructor disappear?

jbsauvan commented on DataFormats/L1THGCal/src/HGCalMulticluster.cc:
```diff
@@ -2,35 +2,76 @@
-    L1Candidate(p4, pt, eta, phi),
-    myclusters_(basic_clusters){
-    } 
+HGCalMulticluster::HGCalMulticluster( const l1t::HGCalCluster & clu )
```
HGCalMultiCluster inherits from L1Candidate. You should make sure that the L1Candidate constructor is properly called.
Also, it is better to initialize the members in the initialization list

jbsauvan commented on DataFormats/L1THGCal/src/HGCalMulticluster.cc:
```diff
@@ -2,35 +2,76 @@
-  else if(hwPt()==cl.hwPt()) {
+  if( mipPt() < cl.mipPt() ) 
+      res = true;
+  else if( mipPt() == cl.mipPt() ) {
```
`mipPt` is a double. The following is not needed

jbsauvan commented on DataFormats/L1THGCal/src/HGCalTriggerCell.cc:
```diff
@@ -2,19 +2,23 @@
-HGCalTriggerCell::
-HGCalTriggerCell( const LorentzVector& p4,
+
+HGCalTriggerCell::HGCalTriggerCell( const LorentzVector& p4,
```
Please don't make useless changes

jbsauvan commented on L1Trigger/L1THGCal/interface/be_algorithms/HGCalClusteringImpl.h:
```diff
@@ -1,24 +1,28 @@
-    HGCalClusteringImpl(const edm::ParameterSet &conf);    
-    void clusterizeBase(const l1t::HGCalTriggerCellBxCollection&, l1t::HGCalClusterBxCollection&);
+    HGCalClusteringImpl( const edm::ParameterSet & conf);    
+    void clusterise( const l1t::HGCalTriggerCellBxCollection & trgcells_, 
```
Not really important, but probably better to have the correct spelling: `clusterize` instead of `clusterise`.

jbsauvan commented on L1Trigger/L1THGCal/interface/be_algorithms/HGCalClusteringImpl.h:
```diff
@@ -1,24 +1,28 @@
     
-    double seed_CUT_;
-    double tc_CUT_;
+    double seedThr_;
```
It's a matter of taste, but I tend to prefer clear member names like:
`seedThreshold_`, `triggerCellThreshold_`
People not familiar with the code should be able to easily understand what these variables are

jbsauvan commented on L1Trigger/L1THGCal/interface/be_algorithms/HGCalMulticlusteringImpl.h:
```diff
@@ -1,22 +1,24 @@
-    void clusterizeMultiple(const l1t::HGCalClusterBxCollection&, l1t::HGCalMulticlusterBxCollection&);
-
+    HGCalMulticlusteringImpl( const edm::ParameterSet &conf);    
+    void clusterise( const l1t::HGCalClusterBxCollection & clusters_, 
```
Same as above. Better to have the correct spelling.

jbsauvan commented on L1Trigger/L1THGCal/interface/be_algorithms/HGCalMulticlusteringImpl.h:
```diff
@@ -1,22 +1,24 @@
     
-    double dR_forC3d_;
+    double dr_;
+    BXVector<edm::PtrVector<l1t::HGCalTriggerCell>> clustersTCcollections_; 
```
What exactly is that BX vector of vectors of trigger cell pointers? I don't see it used anywhere.

jbsauvan commented on L1Trigger/L1THGCal/interface/be_algorithms/HGCalTriggerCellCalibration.h:
```diff
@@ -19,7 +19,8 @@ class HGCalTriggerCellCalibration{
     HGCalTriggerCellCalibration(const edm::ParameterSet &conf);    
-    void calibrate(l1t::HGCalTriggerCell&, int cellThickness); 
+    void calibrateInMipT(l1t::HGCalTriggerCell&, int cellThickness); 
+    void calibrateInGeV(l1t::HGCalTriggerCell&, int cellThickness); 
```
I would like to have as well a method that does MipT -> GeV, e.g.:
`calibrateMipTInGeV` 
And the method `calibrateInGeV` would just call sequentially `calibrateInMipT` and `calibrateMipTInGeV`

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/FullModuleSumAlgo.cc:
```diff
@@ -71,8 +71,8 @@ void FullModuleSumAlgo<FECODEC,DATA>::run(const l1t::HGCFETriggerDigiCollection&
         cluster.setModule(moduleId.wafer());
-        cluster.setLayer(moduleId.layer());
-        cluster.setSubDet(moduleId.subdetId());
+        //cluster.setLayer(moduleId.layer());
```
Since there is the now a seedDetId in the cluster class, you could maybe set it here with the detId of the module

jbsauvan reviewed (commented):
Hi @mastrolorenzo 
Second part of the comments. To be continued...

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -2,10 +2,7 @@
 #include "L1Trigger/L1THGCal/interface/fe_codecs/HGCalTriggerCellBestChoiceCodec.h"
 #include "L1Trigger/L1THGCal/interface/fe_codecs/HGCalTriggerCellThresholdCodec.h"
 #include "DataFormats/ForwardDetId/interface/HGCalDetId.h"
-#include "DataFormats/L1THGCal/interface/HGCalTriggerCell.h"
```
Since you are using collections of trigger cells, clusters and multiclusters, you should keep the three corresponding includes

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -14,119 +11,142 @@ using namespace HGCalTriggerBackend;
-        typedef std::unique_ptr<HGCalTriggerGeometryBase> ReturnType;
-
-        virtual void setProduces(edm::EDProducer& prod) const override final 
+public:
```
Please don't change all the indentations without any reason.

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -14,119 +11,142 @@ using namespace HGCalTriggerBackend;
 
         }
+    
+    virtual void setProduces(edm::EDProducer& prod) const override final 
```
Same thing. Don't change all the indentations without any reason

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -14,119 +11,142 @@ using namespace HGCalTriggerBackend;
+    
+    virtual void setProduces(edm::EDProducer& prod) const override final 
+        {
+            prod.produces<l1t::HGCalTriggerCellBxCollection>( "calibTC" );
```
`calibratedTriggerCells` would be a better collection  name

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -14,119 +11,142 @@ using namespace HGCalTriggerBackend;
         {
-            evt.put(std::move(cluster_product_),name());
-            evt.put(std::move(multicluster_product_),"cluster3D");
+            evt.put( std::move( trgcell_product_ ),      "calibTC"   );
```
Same comment concerning the name of the collection

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -14,119 +11,142 @@ using namespace HGCalTriggerBackend;
+    std::string HGCalEESensitive_;
+    std::string HGCalHESiliconSensitive_;
+    
+    /* handle the collection define form the lables (previous section) */
```
Sorry, but I don't understand this comment

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -14,119 +11,142 @@ using namespace HGCalTriggerBackend;
+    es.get<IdealGeometryRecord>().get( HGCalEESensitive_,        hgceeTopoHandle_ );
+    es.get<IdealGeometryRecord>().get( HGCalHESiliconSensitive_, hgchefTopoHandle_ );
+    
+    std::cout << "Coll size " << coll.size() << std::endl;
```
std::cout to be removed. Or transformed into LogDebug.

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -14,119 +11,142 @@ using namespace HGCalTriggerBackend;
-            {
-
+        
+        for(const auto& triggercell : data.payload){
```
Avoid useless changes.

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -14,119 +11,142 @@ using namespace HGCalTriggerBackend;
+        
+        for(const auto& triggercell : data.payload){
+            
+            if( triggercell.hwPt() > 0 ){
```
ditto

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -14,119 +11,142 @@ using namespace HGCalTriggerBackend;
-                    cellThickness = (hgchefTopoHandle_)->dddConstants().waferTypeL((unsigned int)detid.wafer() );
-                }else if( subdet == HGCHEB ){
+                
+                if( subdet == HGCEE ) 
```
Please put the brackets back. This is dangerous in case someone wants to add a line in the if statement but doesn't notice that the brackets are missing; in that case it creates a bug which is difficult to notice.

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -14,119 +11,142 @@ using namespace HGCalTriggerBackend;
+                
+                if( subdet == HGCEE ) 
+                    cellThickness = hgceeTopoHandle_->dddConstants().waferTypeL( (unsigned int)detid.wafer() );
+                else if( subdet == HGCHEF )
```
ditto

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -14,119 +11,142 @@ using namespace HGCalTriggerBackend;
+                    cellThickness = hgceeTopoHandle_->dddConstants().waferTypeL( (unsigned int)detid.wafer() );
+                else if( subdet == HGCHEF )
+                    cellThickness = hgchefTopoHandle_->dddConstants().waferTypeL( (unsigned int)detid.wafer() );
+                else if( subdet == HGCHEB )
```
ditto

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -14,119 +11,142 @@ using namespace HGCalTriggerBackend;
+                else if( subdet == HGCHEF )
+                    cellThickness = hgchefTopoHandle_->dddConstants().waferTypeL( (unsigned int)detid.wafer() );
+                else if( subdet == HGCHEB )
                     edm::LogWarning("DataNotFound") << "ATTENTION: the BH trgCells are not yet implemented !! ";
```
Probably not necessary to put exclamation points. And write 'trigger cells ' instead of 'trgcells'.

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -14,119 +11,142 @@ using namespace HGCalTriggerBackend;
+    
+                l1t::HGCalTriggerCell calibratedtriggercell( triggercell );
+                calibration_.calibrateInMipT( calibratedtriggercell, cellThickness ); 
+                calibration_.calibrateInGeV( calibratedtriggercell, cellThickness ); 
```
Why do you calibrate the trigger cell twice?

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -14,119 +11,142 @@ using namespace HGCalTriggerBackend;
 DEFINE_EDM_PLUGIN(HGCalTriggerBackendAlgorithmFactory, 
-        HGCClusterAlgoBestChoice,
-        "HGCClusterAlgoBestChoice");
+                  HGCClusterAlgoBestChoice,
```
Avoid useless changes.

jbsauvan replied:
This comment concerning useless changes is still valid

jbsauvan commented on L1Trigger/L1THGCal/plugins/ntuples/HGCalTriggerNtupleHGCClusters.cc:
```diff
@@ -26,6 +26,7 @@ class HGCalTriggerNtupleHGCClusters : public HGCalTriggerNtupleBase
     std::vector<float> cl_energy_;
     std::vector<float> cl_eta_;
     std::vector<float> cl_phi_;
+    std::vector<int> cl_layer_;
```
A cluster is not necessarily a 2D cluster. You can have the layer of the seed trigger cells, But the layer of the cluster doesn't make much sense in the general case

jbsauvan commented on L1Trigger/L1THGCal/python/hgcalTriggerNtuples_cfi.py:
```diff
@@ -39,9 +39,9 @@
-        ntuple_triggercells
-        #  ntuple_clusters,
-        #  ntuple_multicluster
+#        ntuple_digis,
```
Please keep this ntuple in the default parameters

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc:
```diff
@@ -1,52 +1,73 @@
-    tc_CUT_ = conf.getParameter<double>("clustering_threshold");
+HGCalClusteringImpl::HGCalClusteringImpl(const edm::ParameterSet & conf){    
+    
+    seedThr_ = conf.getParameter<double>("seeding_threshold");
```
You could put these initializations in the initialization list of the constructor

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc:
```diff
@@ -1,52 +1,73 @@
+                                      l1t::HGCalClusterBxCollection & clusters_
+    ){
+
+    edm::LogInfo("HGCclusterParameters") << "C2d seeding Thr: " << seedThr_ ; 
```
This is debug information. Change LogInfo to LogDebug.

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc:
```diff
@@ -1,52 +1,73 @@
+
+    /* seeding the TCs */
+    int itc=0;
+    for( l1t::HGCalTriggerCellBxCollection::const_iterator tc = trgcells_.begin(); tc != trgcells_.end(); ++tc,++itc )
```
Add brackets. For the same reason stated in a previous comment.

jbsauvan commented on L1Trigger/L1THGCal/interface/be_algorithms/HGCalClusteringImpl.h:
```diff
@@ -1,24 +1,28 @@
-    void clusterizeBase(const l1t::HGCalTriggerCellBxCollection&, l1t::HGCalClusterBxCollection&);
+    HGCalClusteringImpl( const edm::ParameterSet & conf);    
+    void clusterise( const l1t::HGCalTriggerCellBxCollection & trgcells_, 
+                     l1t::HGCalClusterBxCollection & clusters_ 
```
clusters_ and trgcells_ are not members of the class. Better to remove the trailing underscores.

jbsauvan commented on L1Trigger/L1THGCal/interface/be_algorithms/HGCalMulticlusteringImpl.h:
```diff
@@ -1,22 +1,24 @@
-
+    HGCalMulticlusteringImpl( const edm::ParameterSet &conf);    
+    void clusterise( const l1t::HGCalClusterBxCollection & clusters_, 
+                     l1t::HGCalMulticlusterBxCollection & multiclusters_);
```
clusters_ and multiclusters_ are not members of the class. better to remove the trailing underscores.

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc:
```diff
@@ -1,52 +1,73 @@
+    for( l1t::HGCalTriggerCellBxCollection::const_iterator tc = trgcells_.begin(); tc != trgcells_.end(); ++tc,++itc ){
+
+        if( tc->hwPt() < tcThr_ )
+            continue;
```
Ditto. Put the `continue` on the same line as `if`. Or add brackets.

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc:
```diff
@@ -1,52 +1,73 @@
+        /* searching for TC near the center of the cluster  */
+        int iclu=0;
+        vector<int> tcPertinentClusters; 
+        for( std::vector<l1t::HGCalCluster>::iterator clu = clustersTmp.begin(); clu != clustersTmp.end(); ++clu,++iclu )
```
ditto. This is error prone. Put brackets for the loop and for the if statement.

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc:
```diff
@@ -1,52 +1,73 @@
+                tcPertinentClusters.push_back(iclu);
+
+        if( tcPertinentClusters.size() == 0 && isSeed[itc] ){
+            //edm::PtrVector<l1t::HGCalTriggerCell> coll;
```
Remove commented out code

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc:
```diff
@@ -1,52 +1,73 @@
+            //edm::PtrVector<l1t::HGCalTriggerCell> coll;
+            //clustersTCcollections_.push_back( coll );
+            l1t::HGCalCluster obj( *tc );
+            clustersTmp.push_back( obj );
```
Use `emplace_back` to avoid the copy of the cluster.
```cpp
clustersTmp.emplace_back( *tc );
```

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc:
```diff
@@ -1,52 +1,73 @@
+        }
+        else if ( tcPertinentClusters.size() > 0 ){
+         
+            uint minDist = 300;
```
`uint` is not a c++ standard. Use `unsigned` instead.

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc:
```diff
@@ -1,52 +1,73 @@
+    }
+
+    /* making the collection of clusters */
+    for( unsigned i(0); i<clustersTmp.size(); ++i )
```
Add brackets, for safety reason.

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc:
```diff
@@ -1,52 +1,73 @@
+
+    /* making the collection of clusters */
+    for( unsigned i(0); i<clustersTmp.size(); ++i )
+        clusters_.push_back( 0, clustersTmp.at(i) );
```
Is there a reason why you use a temporary vector of clusters and you copy all of them in the end? Which is maybe not the most efficient thing to do.

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalMulticlusteringImpl.cc:
```diff
@@ -1,69 +1,51 @@
+
+HGCalMulticlusteringImpl::HGCalMulticlusteringImpl( const edm::ParameterSet& config ) 
+{    
+    dr_ = config.getParameter<double>("dR_multicluster");
```
Better to initialize this in the initialization list of the constructor

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalMulticlusteringImpl.cc:
```diff
@@ -1,69 +1,51 @@
+                                           l1t::HGCalMulticlusterBxCollection & multiclusters_)
+{
+   
+    edm::LogInfo("HGCclusterParameters") << "Multicluster dR for Near Neighbour search: " << dr_;  
```
This is more debug information. Change LogInfo to LogDebug.

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalMulticlusteringImpl.cc:
```diff
@@ -1,69 +1,51 @@
-            uint32_t totLayer = 0;
+        int imclu=0;
+        vector<int> tcPertinentMulticlusters;
+        for(l1t::HGCalMulticlusterBxCollection::const_iterator mclu = multiclustersTmp.begin(); mclu != multiclustersTmp.end(); ++mclu,++imclu)
```
Add brackets (for loop and if statement) for safety reason

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalMulticlusteringImpl.cc:
```diff
@@ -1,69 +1,51 @@
-                    totLayer++;
+        if( tcPertinentMulticlusters.size() == 0 ){
+            l1t::HGCalMulticluster obj( *clu );
+            multiclustersTmp.push_back( obj );
```
Probably better to use `emplace_back` to avoid the copy of the cluster.
```cpp
 multiclustersTmp.emplace_back( *clu );
```

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalMulticlusteringImpl.cc:
```diff
@@ -1,69 +1,51 @@
+            multiclustersTmp.push_back( obj );
+        }
+        else{
+            uint minDist = 1;
```
`uint` is not a standard type. Use `unsigned` instead

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalMulticlusteringImpl.cc:
```diff
@@ -1,69 +1,51 @@
+   }
+
+    /* making the collection of multiclusters */
+    for( unsigned i(0); i<multiclustersTmp.size(); ++i )
```
Use brackets for safety reason.

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalMulticlusteringImpl.cc:
```diff
@@ -1,69 +1,51 @@
+
+    /* making the collection of multiclusters */
+    for( unsigned i(0); i<multiclustersTmp.size(); ++i )
+        multiclusters_.push_back( 0, multiclustersTmp.at(i) );
```
Is there a reason why a temporary vector of clusters is used?

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalTriggerCellCalibration.cc:
```diff
@@ -10,31 +10,66 @@ HGCalTriggerCellCalibration::HGCalTriggerCellCalibration(const edm::ParameterSet
-    HGCalDetId trgdetid(trgCell.detId());
-    int trgCellLayer = trgdetid.layer();
+
+void HGCalTriggerCellCalibration::calibrateInMipT(l1t::HGCalTriggerCell& trgCell, int cellThickness)
```
Avoid useless changes

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalTriggerCellCalibration.cc:
```diff
@@ -10,31 +10,66 @@ HGCalTriggerCellCalibration::HGCalTriggerCellCalibration(const edm::ParameterSet
+    int subdet = trgdetid.subdetId();
+
+    /* get the transverse momentum in mip units */
+    //int hwPt = trgCell.hwPt(); /* we will use this one ?*/
```
Is this needed?

jbsauvan reviewed (commented):
Hi @mastrolorenzo 
This is the last set of comments

mastrolorenzo committed:
All minor comments from Jean-Baptiste implemented. Major comments related to the usage of PtrVector (used to get the collection of pertinent trgCell and Cluster) will be included in the next commit

mastrolorenzo committed:
"isPertinent" function has been moved from cluster and multicluster classes to the respective Impl.

mastrolorenzo committed:
1) Addition of Persistent PtrVector to pertinent tc/clusters collections
2) Modification of the classes_def.xml and classes.h
3) Restored the the type of centre and centreProj (previously centreNorm) as GlobalPoint
4) other minor changes + few variable renames + overall cleaning

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
+namespace l1t {
 
-  class HGCalCluster : public L1Candidate {
+    class HGCalCluster : public L1Candidate {
```
Please fix all the indentations that have been changed without reason

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
+        /* constructors and destructor */
+        HGCalCluster(){}
+        HGCalCluster( const LorentzVector p4,
+                      int pt,
```
Keep the default values as it was before

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
-      uint32_t hwPtHad()  const {return hwPtHad_;}
-      uint32_t hwSeedPt() const {return hwSeedPt_;}
+        /* trigger-cells collection pertinent to the cluster */
+        const edm::PtrVector<l1t::HGCalTriggerCell>  triggercells() const { return triggercells_; }        
```
* Better to avoid the copy the vector. return a const reference.
* Also it could be useful to have a method that returns the size of the vector. And methods that return iterators to begin() and end() of the vector.

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
-      uint32_t module()  const {return module_;}
+        /* helpers */
+        void addTriggerCell( const l1t::HGCalTriggerCell &tc );
+        void addTriggerCellList( edm::Ptr<l1t::HGCalTriggerCell> &p );
```
* Why are there two of these addTriggerCell methods? I don't think the first one that takes as input the simple reference is needed. 
* Also you probably want to pass the input parameter with `const`

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
+                      int phi
+            );
+      
+        HGCalCluster( const l1t::HGCalTriggerCell &tc ); 
```
You should pass an edm::Ptr, otherwise you cannot put the seed cluster in the list of clusters.

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
 
-      ClusterShapes shapes ;
+        /* get info */
+        bool isValid()       const { return true;  }
```
Instead of returning 'true', you could check if the cluster contains trigger cells. If not it is not a valid cluster.

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
+        double distance( const l1t::HGCalTriggerCell &tc ) const; /* return distance in 'cm' */
+        
+        GlobalPoint centre() const { return centre_; }
+        GlobalPoint centreProj() const { return centreProj_; }
```
Since you are returning a class member you should better return a const reference to avoid the copy.

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -1,68 +1,95 @@
+        bool isValid()       const { return true;  }
+        double mipPt()       const { return mipPt_; }
+        double seedMipPt()   const { return seedMipPt_; }
+        uint32_t seedDetId() const { return seedDetId_; }
```
It could be useful to add a method that returns the seed trigger cell (the first trigger cell in the vector of trigger cells)

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,78 +1,78 @@
-        typedef edm::PtrVector<l1t::HGCalCluster>  ClusterCollection;
 
-        HGCalMulticluster(){}
+  public:
```
Again, please keep the original indentations unless there is a problem (for instance mixed indentations)

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,78 +1,78 @@
+        HGCalMulticluster() {}
+
         HGCalMulticluster( const LorentzVector p4,
                            int pt,
```
Add default values as in HGCalCluster

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,78 +1,78 @@
-        
+
+        /* cluster collection pertinent to the multicluster */
+        const edm::PtrVector<l1t::HGCalCluster>  clusters() const { return clusters_; }        
```
* Return a const reference to avoid copy
* You could add a size(), begin() and end() methods (same comment as for HGCalCluster)

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,78 +1,78 @@
+
+        /* helpers */
+        void addCluster( const l1t::HGCalCluster & clu );
+        void addClusterList( edm::Ptr<l1t::HGCalCluster> &clu );
```
* Same thing as HGCalCluster, these duplicated methods are probably not needed.
* Also you probably want to pass the input parameter with `const`

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,78 +1,78 @@
+                           int phi
+        );
+
+        HGCalMulticluster( const l1t::HGCalCluster & clu );
```
Pass an edm::Ptr, otherwise you cannot fill the seed cluster in the vector of clusters

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,78 +1,78 @@
-        
-        uint32_t hOverE() const {return hOverE_;}
-        
+        GlobalPoint centre() const { return centre_; } /* in normal plane (x, y, z) */
```
Return const reference to avoid copy

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -1,78 +1,78 @@
+        GlobalPoint centre() const { return centre_; } /* in normal plane (x, y, z) */
+        GlobalPoint centreProj() const { return centreProj_; } /* in normalized plane (x/z, y/z, z/z) */
+
+        uint32_t firstClusterDetId() const { return firstClusterDetId_; }
```
As for HGCalCluster you could in addition have a method that returns the first cluster

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalTriggerCell.h:
```diff
@@ -5,41 +5,57 @@
-            GlobalPoint position_;
+        
+    public:
+        
```
Don't change all the indentations

jbsauvan commented on DataFormats/L1THGCal/src/HGCalCluster.cc:
```diff
@@ -1,34 +1,146 @@
+
+
+HGCalCluster::HGCalCluster(  const l1t::HGCalTriggerCell &tcSeed )
+    : seedDetId_( tcSeed.detId() ),
```
* You should call the constructor of the base class
* And initialize the other class members

jbsauvan commented on DataFormats/L1THGCal/src/HGCalCluster.cc:
```diff
@@ -1,34 +1,146 @@
+   : L1Candidate(p4, pt, eta, phi),
+     centre_(0, 0, 0),
+     mipPt_(0),
+     seedMipPt_(0)
```
You should initialize the other class members as well

jbsauvan commented on DataFormats/L1THGCal/src/HGCalCluster.cc:
```diff
@@ -1,34 +1,146 @@
+                                   tc.position().y(),
+                                   tc.position().z() );
+
+    Basic3DVector<float> centreVector( tc.position().x(), 
```
I think you can give directly the tc position to construct the vectors. It would be probably simpler to read.

jbsauvan commented on DataFormats/L1THGCal/src/HGCalCluster.cc:
```diff
@@ -1,34 +1,146 @@
+                                       tc.position().y(), 
+                                       tc.position().z() );
+
+    centreVector = centreVector*mipPt_ + tcVector*tc.mipPt();
```
I don't understand this line. According to the above `tcVector` and `centreVector` are the same thing. So what exactly is done here?

jbsauvan commented on DataFormats/L1THGCal/src/HGCalCluster.cc:
```diff
@@ -1,34 +1,146 @@
+                          (float)centreVector.y(), 
+                          (float)centreVector.z() );
+
+    GlobalPoint centreProjTmp((float)centreVector.x() / centreVector.z(),  
```
I think you can switch between vector and point directly, like is done in the geometry class [here](https://github.com/PFCal-dev/cmssw/blob/hgc-tpg-devel-CMSSW_9_0_0_pre5/L1Trigger/L1THGCal/plugins/geometries/HGCalTriggerGeometryHexImp2.cc#L355), which is probably easier to read.

jbsauvan commented on DataFormats/L1THGCal/src/HGCalCluster.cc:
```diff
@@ -1,34 +1,146 @@
+                              (float)centreVector.y() / centreVector.z(), 
+                              (float)centreVector.z() / centreVector.z() );
+
+    centre_ = centreTmp;
```
Why creating temporary points and directly copy them into the class members?

jbsauvan commented on DataFormats/L1THGCal/src/HGCalCluster.cc:
```diff
@@ -1,34 +1,146 @@
+    mipPt_ += tc.mipPt();
+
+    int hwPt = 0;
+    hwPt  += tc.hwPt();
```
I think what you want to do is:
```cpp
int hwPt = this->hwPt() + tc.hwPt();
```

jbsauvan commented on DataFormats/L1THGCal/src/HGCalCluster.cc:
```diff
@@ -1,34 +1,146 @@
+double HGCalCluster::distance(const l1t::HGCalTriggerCell &tc) const
+{
+
+    GlobalPoint tcPointXYZ( tc.position().x(), 
```
You don't need that. `tc.position()` is already a GlobalPoint

jbsauvan commented on DataFormats/L1THGCal/src/HGCalCluster.cc:
```diff
@@ -1,34 +1,146 @@
+                             tc.position().y(), 
+                             tc.position().z() );
+    
+    return ( tcPointXYZ - centre_ ).mag();
```
Just write
```cpp
return ( tc.position() - centre_ ).mag();
```

jbsauvan commented on DataFormats/L1THGCal/src/HGCalCluster.cc:
```diff
@@ -1,34 +1,146 @@
+    if( seedDetId.subdetId()==3 ){
+        layer =  seedDetId.layer();
+    }
+    else if( seedDetId.subdetId()==4 ){
```
I would prefer to keep the same definition of layer consistent everywhere (keep the layer of the detid)

jbsauvan commented on DataFormats/L1THGCal/src/HGCalCluster.cc:
```diff
@@ -1,34 +1,146 @@
-  }
-  return res;
+
+    return res;
```
Or you can just return 
```cpp
return (mipPt() < cl.mipPt());
```

jbsauvan commented on DataFormats/L1THGCal/src/HGCalMulticluster.cc:
```diff
@@ -1,36 +1,114 @@
+
+    centreProj_ = clu.centreProj();
+    centre_ = clu.centre();
+    mipPt_ = clu.mipPt();
```
Put the initializations in the initialization  list of the constructor

jbsauvan commented on DataFormats/L1THGCal/src/HGCalMulticluster.cc:
```diff
@@ -1,36 +1,114 @@
 using namespace l1t;
 
+HGCalMulticluster::HGCalMulticluster( const l1t::HGCalCluster & clu )
+    : firstClusterDetId_( clu.seedDetId() )
```
You should add the call to the base class constructor

jbsauvan commented on DataFormats/L1THGCal/src/HGCalMulticluster.cc:
```diff
@@ -1,36 +1,114 @@
+                                       centre_.y(), 
+                                       centre_.z() );
+
+    Basic3DVector<float> cluCentreVector( clu.centre().x(), 
```
Same comment as for HGCalCluster, you can directly construct the vector from the point, which would be easier to read.

jbsauvan commented on DataFormats/L1THGCal/src/HGCalMulticluster.cc:
```diff
@@ -1,36 +1,114 @@
+    GlobalPoint centreAux((float)centreVector.x(), 
+                          (float)centreVector.y(), 
+                          (float)centreVector.z() );
+    centre_ = centreAux;
```
Same as HGCalCluster: why creating a temporary point which is directly copied into the class member?

jbsauvan commented on DataFormats/L1THGCal/src/HGCalMulticluster.cc:
```diff
@@ -1,36 +1,114 @@
+                                           centreProj_.y(),
+                                           centreProj_.z() );
+
+    Basic3DVector<float> cluCentreProjVector( clu.centreProj().x(), 
```
Same comment as above. These vectors can be directly constructed from the point (easier to read).

jbsauvan commented on DataFormats/L1THGCal/src/HGCalMulticluster.cc:
```diff
@@ -1,36 +1,114 @@
+                               (float)centreProjVector.y(), 
+                               (float)centreProjVector.z() );
+
+    centreProj_ = centreProjAux;
```
Same as above concerning the temporary object

jbsauvan commented on DataFormats/L1THGCal/src/HGCalMulticluster.cc:
```diff
@@ -1,36 +1,114 @@
+    mipPt_ += clu.mipPt();
+  
+    int hwPt = 0;
+    hwPt += clu.hwPt();
```
You probably want to do
```cpp
int hwPt = this->hwPt() + clu.hwPt();
```

jbsauvan commented on DataFormats/L1THGCal/src/HGCalMulticluster.cc:
```diff
@@ -1,36 +1,114 @@
+  if( mipPt() < cl.mipPt() ) 
+      res = true;
+
   return res;
```
Simpler to just return
```cpp
return (mipPt() < cl.mipPt());
```

jbsauvan commented on DataFormats/L1THGCal/src/HGCalTriggerCell.cc:
```diff
@@ -2,19 +2,24 @@
 
 using namespace l1t;
 
+
```
This file only has useless changes. Revert back to the original version.

jbsauvan commented on L1Trigger/L1THGCal/interface/HGCalTriggerBackendAlgorithmBase.h:
```diff
@@ -51,9 +51,13 @@ class HGCalTriggerBackendAlgorithmBase {
   //runs the trigger algorithm, storing internally the results
   virtual void setProduces(edm::EDProducer& prod) const = 0;
 
+//  virtual void run(const l1t::HGCFETriggerDigiCollection& coll, 
```
You can remove these commented lines

jbsauvan commented on L1Trigger/L1THGCal/interface/HGCalTriggerBackendProcessor.h:
```diff
@@ -37,14 +37,20 @@ class HGCalTriggerBackendProcessor {
 
   void setProduces(edm::EDProducer& prod) const;
 
+//  void run(const l1t::HGCFETriggerDigiCollection& coll,
```
You can remove these commented lines

jbsauvan commented on L1Trigger/L1THGCal/interface/HGCalTriggerBackendProcessor.h:
```diff
@@ -37,14 +37,20 @@ class HGCalTriggerBackendProcessor {
   void putInEvent(edm::Event& evt);
 
   void reset();
+//  void reset(edm::ConsumesCollector&& cc);
```
You can remove this commented line

jbsauvan commented on L1Trigger/L1THGCal/interface/be_algorithms/HGCalClusteringImpl.h:
```diff
@@ -1,24 +1,31 @@
-    void clusterizeBase(const l1t::HGCalTriggerCellBxCollection&, l1t::HGCalClusterBxCollection&);
+    HGCalClusteringImpl( const edm::ParameterSet & conf);    
+
+    void clusterize( const edm::OrphanHandle<l1t::HGCalTriggerCellBxCollection> triggerCellsHandle,
```
Actually, it might be better to pass an `edm::PtrVector<>`, as done [here](https://cmssdt.cern.ch/lxr/source/RecoLocalCalo/HGCalRecAlgos/interface/HGCalDepthPreClusterer.h#0033)

jbsauvan commented on L1Trigger/L1THGCal/interface/be_algorithms/HGCalClusteringImpl.h:
```diff
@@ -1,24 +1,31 @@
+                     l1t::HGCalClusterBxCollection & clusters 
+        );
+
+    bool isPertinent( const l1t::HGCalTriggerCell &tc, l1t::HGCalCluster &clu, double distXY) const;
```
`clu` is an input parameter. It should be `const`

jbsauvan commented on L1Trigger/L1THGCal/interface/be_algorithms/HGCalMulticlusteringImpl.h:
```diff
@@ -1,22 +1,26 @@
-    void clusterizeMultiple(const l1t::HGCalClusterBxCollection&, l1t::HGCalMulticlusterBxCollection&);
+    HGCalMulticlusteringImpl( const edm::ParameterSet &conf);    
+
+    void clusterize( const edm::OrphanHandle<l1t::HGCalClusterBxCollection> clustersHandle, 
```
Same as above:
Actually, it might be better to pass an `edm::PtrVector<>`

jbsauvan commented on L1Trigger/L1THGCal/interface/be_algorithms/HGCalTriggerCellCalibration.h:
```diff
@@ -19,7 +19,9 @@ class HGCalTriggerCellCalibration{
     HGCalTriggerCellCalibration(const edm::ParameterSet &conf);    
-    void calibrate(l1t::HGCalTriggerCell&, int cellThickness); 
+    void calibrateInMipT(l1t::HGCalTriggerCell&, int cellThickness);
+    void calibrateMipTinGeV(l1t::HGCalTriggerCell&, int cellThickness); 
```
`cellThickness` doesn't seem to be used in this function

jbsauvan commented on L1Trigger/L1THGCal/plugins/HGCalTriggerDigiProducer.cc:
```diff
@@ -23,7 +23,7 @@ class HGCalTriggerDigiProducer : public edm::EDProducer {
   virtual void beginRun(const edm::Run&, 
                         const edm::EventSetup&);
-  virtual void produce(edm::Event&, const edm::EventSetup&);
+    virtual void produce(edm::Event&, const edm::EventSetup&);
```
Fix indentation.

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/FullModuleSumAlgo.cc:
```diff
@@ -71,8 +71,8 @@ void FullModuleSumAlgo<FECODEC,DATA>::run(const l1t::HGCFETriggerDigiCollection&
-        cluster.setLayer(moduleId.layer());
-        cluster.setSubDet(moduleId.subdetId());
+        //cluster.setLayer(moduleId.layer());
+        //cluster.setSubDet(moduleId.subdetId());
```
Remove the commented code

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -16,117 +16,152 @@ class HGCClusterAlgo : public Algorithm<FECODEC>
-            clustering_(conf.getParameterSet("C2d_parameters")),
-            multiclustering_(conf.getParameterSet("C3d_parameters")){}
 
         typedef std::unique_ptr<HGCalTriggerGeometryBase> ReturnType;
```
What is it for?

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -16,117 +16,152 @@ class HGCClusterAlgo : public Algorithm<FECODEC>
         std::unique_ptr<l1t::HGCalMulticlusterBxCollection> multicluster_product_;
+
+        /* orphan handles to the collections of trigger-cells, clusters and multiclusters */
+        edm::OrphanHandle<l1t::HGCalTriggerCellBxCollection> triggerCellsHandle_;
```
These don't need to be class members. They are only used in the `run()` method

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCalSimCluster.cc:
```diff
@@ -151,7 +151,7 @@ namespace HGCalTriggerBackend{
             virtual void run( const l1t::HGCFETriggerDigiCollection & coll,
 		            const edm::EventSetup& es,
-		            const edm::Event&evt
+                    edm::Event&evt
```
Fix indentation

jbsauvan commented on L1Trigger/L1THGCal/plugins/ntuples/HGCalTriggerNtupleHGCClusters.cc:
```diff
@@ -76,6 +79,15 @@ fill(const edm::Event& e, const edm::EventSetup& es)
+    for(unsigned itc(0); itc<pertinentTC.size(); ++itc){
+        nTC++;
+    }
+    cl_ncells_.emplace_back(nTC);
```
You can just do
```cpp
cl_ncells_.emplace_back(cl_itr->triggercells().size())
```
Instead of all your lines above

jbsauvan commented on L1Trigger/L1THGCal/plugins/ntuples/HGCalTriggerNtupleHGCMulticlusters.cc:
```diff
@@ -76,6 +77,14 @@ fill(const edm::Event& e, const edm::EventSetup& es)
+    for(unsigned iclu(0); iclu<pertinentClu.size(); ++iclu){
+        nC2d++;
+    }
+    cl3d_nclu_.emplace_back(nC2d);
```
You can just do
```cpp
cl3d_nclu_.emplace_back(cl3d_itr->clusters().size());
```

jbsauvan commented on L1Trigger/L1THGCal/plugins/ntuples/HGCalTriggerNtupleHGCTriggerCells.cc:
```diff
@@ -3,7 +3,7 @@
 #include "L1Trigger/L1THGCal/interface/HGCalTriggerGeometryBase.h"
 #include "L1Trigger/L1THGCal/interface/HGCalTriggerNtupleBase.h"
-
+#include "DataFormats/Common/interface/OrphanHandle.h"
```
Why do you need that here?

jbsauvan commented on L1Trigger/L1THGCal/src/HGCalTriggerBackendProcessor.cc:
```diff
@@ -26,9 +26,18 @@ void HGCalTriggerBackendProcessor::setProduces(edm::EDProducer& prod) const {
   }
 }
 
+//void HGCalTriggerBackendProcessor::run(const l1t::HGCFETriggerDigiCollection& coll, 
```
You can remove the commented code

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc:
```diff
@@ -1,52 +1,100 @@
+    if( tcDetId.layer() != cluDetId.layer() ||
+        tcDetId.subdetId() != cluDetId.subdetId() ||
+        tcDetId.zside() != cluDetId.zside() )
+        return false;
```
Better to add brackets for safety

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc:
```diff
@@ -1,52 +1,100 @@
+        return false;
+   
+    if ( clu.distance(tc) < distXY )
+        return true;
```
Either write:
```cpp
if ( clu.distance(tc) < distXY ) return true;
```
or 
```cpp
if ( clu.distance(tc) < distXY )
{
  return true;
}
```
If the `if` statement is on a new line, it is safer to add brackets

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc:
```diff
@@ -1,52 +1,100 @@
             }
+            if( tcPertinentClusters.size() == 0 && isSeed[itc] ){
+                l1t::HGCalCluster obj( *tc );
+                clustersTmp.emplace_back( obj );
```
You can directly do
```cpp
clustersTmp.emplace_back(*tc);
``` 
To avoid the copy

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc:
```diff
@@ -1,52 +1,100 @@
-            cluster_product_.push_back(0,cluster);
+
+        /* making the collection of clusters */
+        for( unsigned i(0); i<clustersTmp.size(); ++i ){
```
In principle, it should not be needed to have a temporary collection. It should be possible to fill directly the output collection.

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalMulticlusteringImpl.cc:
```diff
@@ -1,69 +1,76 @@
+            }
+            if( tcPertinentMulticlusters.size() == 0 ){
+                l1t::HGCalMulticluster obj( *clu );
+                multiclustersTmp.emplace_back( obj );
```
You can do directly
```cpp
multiclustersTmp.emplace_back( *clu );
```
To avoid the copy

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalClusteringImpl.cc:
```diff
@@ -1,52 +1,100 @@
+
+        /* making the collection of clusters */
+        for( unsigned i(0); i<clustersTmp.size(); ++i ){
+            if( clustersTmp.size()>0 ){
```
In any case, this is not needed. If you enter the loop it means that the size is >0

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalMulticlusteringImpl.cc:
```diff
@@ -1,69 +1,76 @@
+
+        /* making the collection of multiclusters */
+        for( unsigned i(0); i<multiclustersTmp.size(); ++i ){
+            if( multiclustersTmp.size()>0 ){
```
Same comments as for clusteringImpl

jbsauvan commented on L1Trigger/L1THGCal/src/be_algorithms/HGCalTriggerCellCalibration.cc:
```diff
@@ -10,31 +10,74 @@ HGCalTriggerCellCalibration::HGCalTriggerCellCalibration(const edm::ParameterSet
+    double trgCellMipPt = trgCellMipP/cosh( trgCell.eta() ); 
+
+    /* setting pT [mip] */
+    trgCell.setMipPt( trgCellMipPt ) ;
```
Maybe you could as well set the 4-vector in mips

jbsauvan commented on L1Trigger/L1THGCal/test/testNtuplyzer_cfg.py:
```diff
@@ -25,12 +25,19 @@
+    secondaryFileNames = cms.untracked.vstring(),
+    fileNames = cms.untracked.vstring(
+#        'file:./testGenOnly.root'
+        'file:./test.root'
```
Don't use local files

jbsauvan reviewed (commented):
In addition to the other comments, the file `testHGCClustering.cc_old` should be removed

jbsauvan commented:
Hi @mastrolorenzo 
Could you fix the other test configs. They are not working anymore.
Thanks

mastrolorenzo commented:
Hi JB,

     thanks for the comments, I will also update the other test configs, 
however I have few remarks/question on a couple of comments:

1) In DataFormats/L1THGCal/src/HGCalCluster.cc 
<https://github.com/PFCal-dev/cmssw/pull/87/files/dd198b357b68a5c8f2673da3a74e905af5aaa256#diff-e6db97af4c224e332597e33dbe216694>: 
I would prefer to keep the same definition of layer consistent 
everywhere (keep the layer of the detid)

==> in that case we would need in future steps something that will do 
the jobs to assign to layer[0,12] belonging to subdetid=FH the 
correspondent layer number [28,40], otherwise we will have a 
superposition in the first 12 layer;

2) In L1Trigger/L1THGCal/interface/be_algorithms/HGCalClusteringImpl.h 
<https://github.com/PFCal-dev/cmssw/pull/87/files/dd198b357b68a5c8f2673da3a74e905af5aaa256#diff-7478806ddf0c695019e175352d6ed13d>: 
Actually, it might be better to pass an |edm::PtrVector<>|, as done here 
<https://cmssdt.cern.ch/lxr/source/RecoLocalCalo/HGCalRecAlgos/interface/HGCalDepthPreClusterer.h#0033>

==> this implies that I should build the edm::PtrVector in the 
HGCCluster class, so do yuo know how to pass from handle to PtrVector 
direcly? otherwise I should do as I do in the HGCalClusteringImpl.cc and 
take from a loop over the handle the Ptr<>  and store the Ptr in a 
PtrVector.. but maybe there is a wasy to extract the PtrVector direclty 
from the handle;

3) In 
<https://github.com/PFCal-dev/cmssw/pull/87/files/dd198b357b68a5c8f2673da3a74e905af5aaa256#diff-e5dbebcc3b182bb110e97e84414465be>DataFormats/L1THGCal/interface/HGCalCluster.h 
<https://github.com/PFCal-dev/cmssw/pull/87/files/dd198b357b68a5c8f2673da3a74e905af5aaa256#diff-96b1af9d6cb5ebd269e9f8203581c65e>: 
Why are there two of these addTriggerCell methods? I don't think the 
first one that takes as input the simple reference is needed. 
<https://github.com/PFCal-dev/cmssw/pull/87/files/dd198b357b68a5c8f2673da3a74e905af5aaa256#diff-e5dbebcc3b182bb110e97e84414465be>

==>Yes I know it will be clearer, I will try to do that... my wander is 
I have to change the parameter of all the functions (since thay take so 
far a simple reference to the const iterator, and I will feed the 
addTriggerCellList() function instead with a Ptr (this is if we would 
like to unify the function addTriggerCell() and addTriggerCellList(), so 
they keep a unique input argument, and must be the Ptr to the tc). 
<https://github.com/PFCal-dev/cmssw/pull/87/files/dd198b357b68a5c8f2673da3a74e905af5aaa256#diff-e5dbebcc3b182bb110e97e84414465be>

All the other comments are clear to me and I am going to implement them,

cheers,

Luca


Il 10/03/2017 13:10, Jean-Baptiste Sauvan ha scritto:
>
> Hi @mastrolorenzo <https://github.com/mastrolorenzo>
> Could you fix the other test configs. They are not working anymore.
> Thanks
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/PFCal-dev/cmssw/pull/87#issuecomment-285653854>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AFS0Zwi2xc3iOuWFifdyGmD79pC1JJgSks5rkT28gaJpZM4MTDFZ>.
>

-- 
---------------------------------------------------------------
Dr. Luca Mastrolorenzo
CERN - CMS experiment
Blackett Laboratory - Imperial College London
Prince Consort Rd, London SW7 2BB
United Kingdom

Centre Européen pour la Recherche Nucléaire
Route de Meyrin 385, 1217 Meyrin, Suisse

Office: 40-4B-08
Téléphone: +44 777 8 224 851
---------------------------------------------------------------

jbsauvan commented:
1) 
> in that case we would need in future steps something that will do
the jobs to assign to layer[0,12] belonging to subdetid=FH the
correspondent layer number [28,40], otherwise we will have a
superposition in the first 12 layer;

Do you have specific examples where this would be needed?

---
2)
> this implies that I should build the edm::PtrVector in the
HGCCluster class, so do yuo know how to pass from handle to PtrVector
direcly? otherwise I should do as I do in the HGCalClusteringImpl.cc and
take from a loop over the handle the Ptr<>  and store the Ptr in a
PtrVector.. but maybe there is a wasy to extract the PtrVector direclty
from the handle;

The PtrVector could be built in the HGCClusterAlgo. I don't think you can go directly from collection Handle to PtrVector. But you can do as done [here](https://cmssdt.cern.ch/lxr/source/RecoLocalCalo/HGCalRecProducers/plugins/HGCalClusterTestProducer.cc#0174).

---
3) 
I would just change the `addTriggerCell()` method to
```cpp
void HGCalCluster::addTriggerCell(const edm::Ptr<l1t::HGCalTriggerCell> &p)
```
And remove the `addTriggerCellList()` method

mastrolorenzo committed:
Update Jean-Baptiste comments of the 10/03/17, in particluar, the handles in input to the clusterization functions have been replaced by PtrVectors

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -12,42 +12,56 @@
         /* trigger-cells collection pertinent to the cluster */
-        const edm::PtrVector<l1t::HGCalTriggerCell>  triggercells() const { return triggercells_; }        
+        const edm::PtrVector<l1t::HGCalTriggerCell>& triggercells() const {
+            return (edm::PtrVector<l1t::HGCalTriggerCell>&) triggercells_; 
```
It is not needed to convert to reference

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -12,42 +12,56 @@
         
-        GlobalPoint centre() const { return centre_; }
-        GlobalPoint centreProj() const { return centreProj_; }
+        GlobalPoint& centre() const { return (GlobalPoint&) centre_; }
```
* return const reference
* Conversion to reference not needed

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -12,42 +12,56 @@
-        GlobalPoint centre() const { return centre_; }
-        GlobalPoint centreProj() const { return centreProj_; }
+        GlobalPoint& centre() const { return (GlobalPoint&) centre_; }
+        GlobalPoint& centreProj() const { return (GlobalPoint&) centreProj_; }
```
ditto

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -10,38 +10,53 @@
-        const edm::PtrVector<l1t::HGCalCluster>  clusters() const { return clusters_; }        
-
+        const edm::PtrVector<l1t::HGCalCluster>&  clusters() const {
+            return (edm::PtrVector<l1t::HGCalCluster>&) clusters_; 
```
Conversion to reference not needed

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -10,38 +10,53 @@
+            return false;
+        }
+
+        GlobalPoint& centre() const { return (GlobalPoint&) centre_; }         /* in (x, y, z) */
```
* Return const reference
* Conversion  to reference not needed

mastrolorenzo committed:
fix indentation

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -10,38 +10,53 @@
+        }
+
+        GlobalPoint& centre() const { return (GlobalPoint&) centre_; }         /* in (x, y, z) */
+        GlobalPoint& centreProj() const { return (GlobalPoint&) centreProj_; } /* in (x/z, y/z, z/z) */
```
ditto

jbsauvan commented on DataFormats/L1THGCal/src/HGCalCluster.cc:
```diff
@@ -32,65 +35,42 @@ HGCalCluster::~HGCalCluster()
 
 
-double HGCalCluster::distance(const l1t::HGCalTriggerCell &tc) const
+double HGCalCluster::distance(const edm::Ptr<l1t::HGCalTriggerCell> tc) const
```
It was fine before. An edm::Ptr is not needed here, so it is better to have more standard function arguments

jbsauvan commented on L1Trigger/L1THGCal/interface/be_algorithms/HGCalClusteringImpl.h:
```diff
@@ -14,11 +14,11 @@ class HGCalClusteringImpl{
     HGCalClusteringImpl( const edm::ParameterSet & conf);    
 
-    void clusterize( const edm::OrphanHandle<l1t::HGCalTriggerCellBxCollection> triggerCellsHandle,
+    void clusterize( const edm::PtrVector<l1t::HGCalTriggerCell> triggerCellsPtrs,
```
Use const reference to PtrVector

jbsauvan commented on L1Trigger/L1THGCal/interface/be_algorithms/HGCalClusteringImpl.h:
```diff
@@ -14,11 +14,11 @@ class HGCalClusteringImpl{
         );
 
-    bool isPertinent( const l1t::HGCalTriggerCell &tc, l1t::HGCalCluster &clu, double distXY) const;
+    bool isPertinent( const edm::Ptr<l1t::HGCalTriggerCell> tc, const l1t::HGCalCluster &clu, double distXY) const;
```
The first parameter was fine before (`const  l1t::HGCalTriggerCell &tc`) . It is not needed to have an edm::Ptr here, so standard arguments should be preferred.

jbsauvan commented on L1Trigger/L1THGCal/interface/be_algorithms/HGCalMulticlusteringImpl.h:
```diff
@@ -12,10 +12,10 @@ class HGCalMulticlusteringImpl{
                      l1t::HGCalMulticlusterBxCollection & multiclusters);
  
-    bool isPertinent( const l1t::HGCalCluster & clu, const l1t::HGCalMulticluster & mclu, double dR ) const;
+    bool isPertinent( const edm::Ptr<l1t::HGCalCluster> clu, const l1t::HGCalMulticluster & mclu, double dR ) const;
```
The const reference to HGCalCluster was fine before. Keep it.

jbsauvan reviewed (commented):
Remaining stuff

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -16,113 +16,159 @@ class HGCClusterAlgo : public Algorithm<FECODEC>
-        virtual void run(const l1t::HGCFETriggerDigiCollection& coll, const edm::EventSetup& es, const edm::Event&evt ) override final;
-        virtual void putInEvent(edm::Event& evt) override final 
+    
+    virtual void setProduces(edm::EDProducer& prod) const override final
```
Remaining indentation problem

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -16,113 +16,159 @@ class HGCClusterAlgo : public Algorithm<FECODEC>
         }
+    
+    
+    virtual void run(const l1t::HGCFETriggerDigiCollection& coll, const edm::EventSetup& es, edm::Event&evt ) override final;
```
Remaining indentation problem

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -16,113 +16,159 @@ class HGCClusterAlgo : public Algorithm<FECODEC>
+
 
-        virtual void reset() override final 
+    virtual void putInEvent(edm::Event& evt) override final 
```
Remaining indentation problem

jbsauvan commented on L1Trigger/L1THGCal/plugins/be_algorithms/HGCClusterAlgo.cc:
```diff
@@ -16,113 +16,159 @@ class HGCClusterAlgo : public Algorithm<FECODEC>
         }
+    
 
+    virtual void reset() override final 
```
Remaining indentation problem

jbsauvan reviewed (commented):
Remaining indentation problems

mastrolorenzo committed:
fix few remaining comments + HGCClusterAlgo.cc indentation

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -59,10 +59,10 @@ namespace l1t {
       
-      GlobalPoint& centre() const { return (GlobalPoint&) centre_; }
-      GlobalPoint& centreProj() const { return (GlobalPoint&) centreProj_; }
+      GlobalPoint centre() const { return (const GlobalPoint&) centre_; }
```
It should be:
```cpp
const GlobalPoint& centre() const { return centre_; }
```

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalCluster.h:
```diff
@@ -59,10 +59,10 @@ namespace l1t {
-      GlobalPoint& centre() const { return (GlobalPoint&) centre_; }
-      GlobalPoint& centreProj() const { return (GlobalPoint&) centreProj_; }
+      GlobalPoint centre() const { return (const GlobalPoint&) centre_; }
+      GlobalPoint centreProj() const { return (const GlobalPoint&) centreProj_; }
```
It should be 
```cpp
const GlobalPoint& centreProj() const { return centreProj_; }
```

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -51,8 +51,8 @@ namespace l1t {
 
-        GlobalPoint& centre() const { return (GlobalPoint&) centre_; }         /* in (x, y, z) */
-        GlobalPoint& centreProj() const { return (GlobalPoint&) centreProj_; } /* in (x/z, y/z, z/z) */
+        GlobalPoint centre() const { return (const GlobalPoint&) centre_; }         /* in (x, y, z) */
```
same as above

jbsauvan commented on DataFormats/L1THGCal/interface/HGCalMulticluster.h:
```diff
@@ -51,8 +51,8 @@ namespace l1t {
-        GlobalPoint& centre() const { return (GlobalPoint&) centre_; }         /* in (x, y, z) */
-        GlobalPoint& centreProj() const { return (GlobalPoint&) centreProj_; } /* in (x/z, y/z, z/z) */
+        GlobalPoint centre() const { return (const GlobalPoint&) centre_; }         /* in (x, y, z) */
+        GlobalPoint centreProj() const { return (const GlobalPoint&) centreProj_; } /* in (x/z, y/z, z/z) */
```
same as above

jbsauvan reviewed (commented):
Also, the test configs like `testHGCalL1T_RelVal_cfg.py`  are still not working. Please fix them.

mastrolorenzo committed:
update of the configuration files

mastrolorenzo commented:
Hi JB,

     I fixed the centre's const refs and all the config files that make 
use of the ntuple_step in their sequence (all blue files below).

However, there are other configs that complain because different reasons 
(I report below the cmssw exceptions I got for each of the related files 
(in red) run, but they seems not related to the additions/modifications 
made in the last weeks||
||

|- [x] testHGCalL1T_reproduce_cfg.py
- [x] testHGCalL1T_cfg.py
- [x] testHGCalL1TBestChoiceTriggerCell_cfg.py
- [x] L1TC_CalibWeightExt_cfg.py
- [x] L1Clustering_cfg.py
- [x] testNtuplyzer_cfg.py
- [x] testHGCalL1T_RelVal_cfg.py
- [x] testHGCalL1TThresholdTriggerCell_cfg.py

- [ ] *testHGCalL1TGeometry_cfg.py*
----- Begin Fatal Exception 14-Mar-2017 17:11:40 CET-----------------------
An exception of category 'FileFlushError' occurred while
    [0] Calling File::flush()
Exception Message:
fdatasync() failed with system error 'No such file or directory' (error 
code 2)
----- End Fatal Exception -------------------------------------------------



- [ ] *produceHGCalBestChoiceMonitorNtuple_cfg.py
*----- Begin Fatal Exception 14-Mar-2017 17:59:24 CET-----------------------
An exception of category 'ConfigFileReadError' occurred while
    [0] Processing the python configuration file named 
produceHGCalBestChoiceMonitorNtuple_cfg.py
Exception Message:
python encountered the error: <type 'exceptions.ImportError'>
No module named GeometryExtended2023D3Reco_cff
----- End Fatal Exception -------------------------------------------------
|

|In this file I also tried to update the geometry from D3 to D4 and 
remove the process=cust_2023tilted(process) but in that case I got the 
exception below:|

||

|
----- Begin Fatal Exception 14-Mar-2017 17:07:26 CET-----------------------
An exception of category 'StdException' occurred while
    [0] Processing run: 1
    [1] Calling global beginRun for module 
HGCalTriggerBestChoiceMonitor/'hgcaltriggerntuplizer'
Exception Message:
A std::exception was thrown.
map::at
----- End Fatal Exception -------------------------------------------------
terminate called after throwing an instance of 'edm::Exception'
   what():  An exception of category 'LogicError' occurred.
Exception Message:
Geant4 is still alive, master thread state must be set to EndRun before 
Destruct



- [ ] *testHGCalL1TBestChoice_cfg.py*
----- Begin Fatal Exception 14-Mar-2017 17:03:38 CET-----------------------
An exception of category 'PluginNotFound' occurred while
    [0] Constructing the EventProcessor
    [1] Constructing module: class=HGCalTriggerDigiFEReproducer 
label='hgcalTriggerPrimitiveDigiFEReproducer'
Exception Message:
Unable to find plugin 'SingleCellClusterAlgo' in category 
'HGCalTriggerBackendAlgorithmFactory'. Please check spelling of name.
----- End Fatal Exception -------------------------------------------------
|
||

|I committed/pushed the changes for all the fixed configs.. what we do 
with this three files?|

|cheers,
|

|Luca|

||

|
|

Il 14/03/2017 15:00, Jean-Baptiste Sauvan ha scritto:
>
> *@jbsauvan* commented on this pull request.
>
> Also, the test configs like |testHGCalL1T_RelVal_cfg.py| are still not 
> working. Please fix them.
>
> ------------------------------------------------------------------------
>
> In DataFormats/L1THGCal/interface/HGCalCluster.h 
> <https://github.com/PFCal-dev/cmssw/pull/87#discussion_r105925301>:
>
> >
> -      GlobalPoint& centre() const { return (GlobalPoint&) centre_; }
> -      GlobalPoint& centreProj() const { return (GlobalPoint&) centreProj_; }
> +      GlobalPoint centre() const { return (const GlobalPoint&) centre_; }
>
> It should be:
>
> const  GlobalPoint&centre()const  {return  centre_; }
> ------------------------------------------------------------------------
>
> In DataFormats/L1THGCal/interface/HGCalCluster.h 
> <https://github.com/PFCal-dev/cmssw/pull/87#discussion_r105925482>:
>
> >
> -      GlobalPoint& centre() const { return (GlobalPoint&) centre_; }
> -      GlobalPoint& centreProj() const { return (GlobalPoint&) centreProj_; }
> +      GlobalPoint centre() const { return (const GlobalPoint&) centre_; }
> +      GlobalPoint centreProj() const { return (const GlobalPoint&) centreProj_; }
>
> It should be
>
> const  GlobalPoint&centreProj()const  {return  centreProj_; }
> ------------------------------------------------------------------------
>
> In DataFormats/L1THGCal/interface/HGCalMulticluster.h 
> <https://github.com/PFCal-dev/cmssw/pull/87#discussion_r105925619>:
>
> > @@ -51,8 +51,8 @@ namespace l1t {
>               return false;
>           }
>   
> -        GlobalPoint& centre() const { return (GlobalPoint&) centre_; }         /* in (x, y, z) */
> -        GlobalPoint& centreProj() const { return (GlobalPoint&) centreProj_; } /* in (x/z, y/z, z/z) */
> +        GlobalPoint centre() const { return (const GlobalPoint&) centre_; }         /* in (x, y, z) */
>
> same as above
>
> ------------------------------------------------------------------------
>
> In DataFormats/L1THGCal/interface/HGCalMulticluster.h 
> <https://github.com/PFCal-dev/cmssw/pull/87#discussion_r105925684>:
>
> > @@ -51,8 +51,8 @@ namespace l1t {
>               return false;
>           }
>   
> -        GlobalPoint& centre() const { return (GlobalPoint&) centre_; }         /* in (x, y, z) */
> -        GlobalPoint& centreProj() const { return (GlobalPoint&) centreProj_; } /* in (x/z, y/z, z/z) */
> +        GlobalPoint centre() const { return (const GlobalPoint&) centre_; }         /* in (x, y, z) */
> +        GlobalPoint centreProj() const { return (const GlobalPoint&) centreProj_; } /* in (x/z, y/z, z/z) */
>
> same as above
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/PFCal-dev/cmssw/pull/87#pullrequestreview-26825419>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AFS0Zzt2hfyNDfbYB4JgwZF6em6_PcjZks5rlquTgaJpZM4MTDFZ>.
>

-- 
---------------------------------------------------------------
Dr. Luca Mastrolorenzo
CERN - CMS experiment
Blackett Laboratory - Imperial College London
Prince Consort Rd, London SW7 2BB
United Kingdom

Centre Européen pour la Recherche Nucléaire
Route de Meyrin 385, 1217 Meyrin, Suisse

Office: 40-4B-08
Téléphone: +44 777 8 224 851
---------------------------------------------------------------

jbsauvan commented:
Thanks @mastrolorenzo 
The configs `produceHGCalBestChoiceMonitorNtuple_cfg.py` and `testHGCalL1TBestChoice_cfg.py` have not been updated since a long time. I should do some cleaning.
The geometry test config should work, but I will check later.

jbsauvan commented:
+1
